src/mappings/helper.ts

```
@@ -1,4 +1,41 @@
 import { CosmosEvent, CosmosMessage } from "@subql/types-cosmos";
+import { Message } from "../types";
 
 export const messageId = (msg: CosmosMessage | CosmosEvent): string =>
     `${msg.tx.hash}-${msg.idx}`;
+
+export const referenceEntityInMessage = async (
+    msg: CosmosMessage,
+    entity: {
+        id: string;
+        messageField: keyof Pick<
+            Message,
+            | "objectariumObjectId"
+            | "objectariumId"
+            | "blockId"
+            | "transactionId"
+        >;
+    },
+): Promise<void> => {
+    const message = await Message.get(messageId(msg));
+    if (!message) {
+        return;
+    }
+
+    switch (entity.messageField) {
+        case "objectariumObjectId":
+            message.objectariumObjectId = entity.id;
+            break;
+        case "objectariumId":
+            message.objectariumId = entity.id;
+            break;
+        case "blockId":
+            message.blockId = entity.id;
+            break;
+        case "transactionId":
+            message.transactionId = entity.id;
+            break;
+    }
+
+    await message.save();
```

Just for my information, the `save` method never return an error if the operation fails?

---

```ts
    async save(): Promise<void>
```

Raises an exception in case of failure, I presume.

---

So, I would have expected at least to be wrapped into a `try / catch` block.

---

What kind of processing do you want to do in the catch? There's nothing we can do but let the exception propagate through the promise, imo.

---

If the error is not well handled at top level, you will lose all runtime context, IMHO

---

Sure. I see your point. But this system is quite different than the usual ones.

For the actual cases, our mapping function has less contextual information compared to the runner, which operates at both the top and bottom levels (in sandwitch). Given that we're operating in a sandboxed environment, there's limited additional context we can provide. I'd also like to add that since the processing are deterministic, which means they can be re-played, the upstream context is totally known.

So, for your concern: is the subql runner doing its job? Let's see. To be sure, I introduced a failure in the code to see. And it looks like we have a pretty good context in the logs provided by subquery. For example:

```
subql-okp4-subquery-node-1   | 2023-09-06T15:28:36.996Z <BlockDispatcherService> ERROR failed to index block at height 2720090 handleTransaction(AssertionError [ERR_ASSERTION]: id must be provided
...
subql-okp4-subquery-node-1   |     at t.Block.create (webpack://@okp4/subql-okp4/src/types/models/Block.ts:44:51)
subql-okp4-subquery-node-1   |     at t.handleBlock (webpack://@okp4/subql-okp4/src/mappings/blockHandlers.ts:11:24)
...
subql-okp4-subquery-node-1   |   code: 'ERR_ASSERTION',
subql-okp4-subquery-node-1   |   actual: false,
subql-okp4-subquery-node-1   |   expected: true,
subql-okp4-subquery-node-1   |   operator: '==',
subql-okp4-subquery-node-1   |   handler: 'handleTransaction'
subql-okp4-subquery-node-1   | }
```

Not able to test all the cases, but for I/O erros, in the code there are also pretty well managed (see [store.service.ts](https://github.com/subquery/subql/blob/8a64ae4cb705d6d9100688eda5f13b751a14d603/packages/node-core/src/indexer/store.service.ts#L891) for instance).

I hope this sheds some light on this concern, which is really important for operationality, thanks for raising this point. Note that I have nothing against this approach, but my fear is to add a lot of code whose usefulness appears very low to me.

---

@ccamel 
Thanks for this detailed explanation, if errors are well handled at top level, there is no necessary to add more context.